NIFI-2142 Cache compiled XSLT in TransformXml

jfrazee opened the pull request:
Two other tidbits:

1. This also modifies the XSLT_FILENAME attribute to support expression language.
2. This updated the test files to have 3 space indents because Saxon defaults to that and AFAICT the free version doesn't let you change that -- it's either no indent or 3 space. Could be wrong though.

See https://github.com/apache/nifi/pull/591 for PR against 0.x and past discussion.

JPercivall commented:
Hey @jfrazee, looks like you have contrib check issues that cause the build to fail:

[WARNING] src/main/java/org/apache/nifi/processors/standard/TransformXml.java[24:8] (imports) UnusedImports: Unused import - java.util.Collection.
[WARNING] src/test/java/org/apache/nifi/processors/standard/TestTransformXml.java[19] (imports) AvoidStarImport: Using the '.*' form of import should be avoided - java.io.*.
[WARNING] src/test/java/org/apache/nifi/processors/standard/TestTransformXml.java[36:8] (imports) UnusedImports: Unused import - org.junit.Ignore.

jfrazee commented:
@JPercivall Oh, right. I think these are corrected now but just noticed the image bundle has a bunch of license violations (even after rebasing).

JPercivall commented:
@jfrazee the image bundle is no longer in master (replaced by the media bundle). Do a "git clean -i -d" to remove any old bundles.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TransformXml.java:
```diff
@@ -76,25 +84,59 @@
+            .description("Maximum size of the stylesheet cache.")
+            .required(true)
+            .defaultValue("100")
+            .addValidator(StandardValidators.NON_NEGATIVE_INTEGER_VALIDATOR)
```
Is this property a max character length or data size limit?

JPercivall replied:
Also, this is a integer validator but the property is interpreted as a long on line 171

JPercivall replied:
Actually my new guess is that this is the maximum number of stylesheets stored (the description could be added to, lol)

jfrazee replied:
@JPercivall Yes, that's correct. So "Maximum number of stylesheets to store in the cache"?

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TransformXml.java:
```diff
@@ -76,25 +84,59 @@
+    public static final PropertyDescriptor CACHE_DURATION = new PropertyDescriptor.Builder()
+            .name("cache-duration")
+            .displayName("Cache duration")
+            .description("How long to keep stylesheets in the cache.")
```
This property is misleading. What I initially thought it to be was how long the stylesheets will exist in the cache (deleted after 60s). What it actually is, is an expiration time where after 60s of it not being used it will be removed.

jfrazee replied:
@JPercivall It can be either. Do you have thoughts on which is preferable. If kept as is, how does "How long to keep stylesheets in the cache after last time of access" sound?

JPercivall commented:
@jfrazee I've made a couple comments. Looks good overall, mainly just documentation comments. Do you by chance have a template and/or some example xml/XSLT that I could use for testing to verify the changes?

jfrazee commented:
1. @JPercivall I addressed all your comments above and added a template to https://gist.github.com/jfrazee/a501336826aa77d0daa74e8e0aa3afd0.
2. There is one change in behavior now -- a cache size of 0 disables the cache whereas it was just allowing for an arbitrarily large cache. 
3. There's a screenshot of the example data flow added to https://issues.apache.org/jira/browse/NIFI-2142 which illustrates the difference between running with and without caching. It's not 100% apples to apples with the older TransformXml, but behaves the same way. The queues back up and there's a stark contrast in Tasks/Time.
4. I rebased on the latest master, but contrib-check is failing because of the commenting style on AbstractSNMPProcessor.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TransformXml.java:
```diff
@@ -168,15 +168,18 @@ protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String
+        CacheBuilder cacheBuilder = CacheBuilder.newBuilder().maximumSize(cacheSize);
+
+        if (cacheSize <= 0) {
+            logger.warn("Stylesheet cache disabled because cache size is set to 0");
```
The cache will still be created and on line 213 it will try to use it to look up an xslt if the cache size is 0. I'd suggest not even creating it in that case.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TransformXml.java:
```diff
@@ -101,16 +101,16 @@
+    public static final PropertyDescriptor CACHE_TTL_AFTER_LAST_ACCESS = new PropertyDescriptor.Builder()
+            .name("cache-ttl-after-last-access")
+            .displayName("Cache TTL after last access")
+            .description("How long to keep stylesheets in the cache after last access.")
```
I like using TTL in the display name in order to make it more compact but not every user may know what TTL means so it'd nice to have the explanation in the description.

JPercivall commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TransformXml.java:
```diff
@@ -101,16 +101,16 @@
-            .description("Maximum size of the stylesheet cache.")
+            .description("Maximum number of stylesheets to cache. Zero disables the cache.")
             .required(true)
             .defaultValue("100")
```
This seems super high as a default number of stylesheets. I don't have much experience with xml and stylesheets but I would think that the average user of a default value wouldn't need to cache use more than 10. I will defer to your experience with the xml and stylesheets if you think otherwise though.

jfrazee commented:
@JPercivall Addressed your latest comments and added a test with the cache disabled since the logic varies depending on that value now.

JPercivall commented:
Sorry for taking so long @jfrazee, I got caught up with something I thought was an issue but turned out it was just me messing up. 

+1

Visually verified code, and did a contrib check build. In standalone and a secure 3 node cluster tested the caching using the provided template. Even with just 5 stylesheets being used I saw speed improvements of ~10x! Thanks @jfrazee, I will merge it in.
